File: playwright-java/src/test/java/testtracks/steps/restfulbooker/api/room/RoomApi.java
package testtracks.steps.restfulbooker.api.room;

import com.fasterxml.jackson.databind.node.ObjectNode;
import com.microsoft.playwright.APIRequestContext;
import java.util.HashMap;
import java.util.Map;
import testtracks.api.common.BaseApi;
import testtracks.configuration.TestSettings;
import testtracks.data.JsonPayloads;
import testtracks.diagnostics.ApiEvidence;

public final class RoomApi extends BaseApi {
  public RoomApi(TestSettings settings, APIRequestContext request, ApiEvidence evidence) {
    super(settings, request, evidence);
  }

  public Room createRoom(ObjectNode payload, String token) {
    ensureOk("POST", "room", payload, headersWithToken(token, null), null);

    var rooms = getRooms(null, null);
    var roomName = JsonPayloads.stringValue(payload, "roomName");
    return rooms.items().stream()
        .filter(candidate -> candidate.roomName().equals(roomName))
        .reduce((first, second) -> second)
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "Created room '" + roomName + "' was not returned by GET room."));
  }

  public Room getRoom(int roomId) {
    return sendJson("GET", "room/" + roomId, null, null, null, Room.class);
  }

  public Rooms getRooms(Map<String, String> headers, Map<String, String> query) {
    return sendJson("GET", "room", null, headers, query, Rooms.class);
  }

  public void deleteRoom(int roomId, String token) {
    ensureOk("DELETE", "room/" + roomId, null, headersWithToken(token, null), null);
  }

  private static Map<String, String> headersWithToken(String token, Map<String, String> headers) {
    var merged = headers == null ? new HashMap<String, String>() : new HashMap<>(headers);
    merged.put("Cookie", "token=" + token);
    return merged;
  }
}

File: playwright-csharp/tests/Steps/RestfulBooker/API/Room/RoomApi.cs
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using Microsoft.Playwright;
using TestTracks.Playwright.CSharp.API.Common;
using TestTracks.Playwright.CSharp.Configuration;
using TestTracks.Playwright.CSharp.Diagnostics;

namespace TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Room;

public sealed class RoomApi : BaseApi
{
    public RoomApi(TestSettings settings, IAPIRequestContext request, ApiEvidence? evidence = null)
        : base(settings, request, evidence)
    {
    }

    public async Task<Room> CreateRoom(
        JsonObject payload,
        string token,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        await EnsureOkAsync("POST", "room", payload, HeadersWithToken(token, headers), query);

        var rooms = await GetRooms(headers: headers);
        var roomName = PayloadString(payload, "roomName");
        return rooms.Items.LastOrDefault(candidate => candidate.RoomName == roomName)
            ?? throw new InvalidOperationException($"Created room '{roomName}' was not returned by GET room.");
    }

    public async Task<Room> GetRoom(
        int roomId,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        return await SendJsonAsync<Room>("GET", $"room/{roomId}", headers: headers, query: query);
    }

    public async Task<Rooms> GetRooms(
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        return await SendJsonAsync<Rooms>("GET", "room", headers: headers, query: query);
    }

    public async Task DeleteRoom(
        int roomId,
        string token,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        await EnsureOkAsync("DELETE", $"room/{roomId}", headers: HeadersWithToken(token, headers), query: query);
    }

    private static IReadOnlyDictionary<string, string> HeadersWithToken(
        string token,
        IReadOnlyDictionary<string, string>? headers = null)
    {
        var merged = headers is null
            ? new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            : new Dictionary<string, string>(headers, StringComparer.OrdinalIgnoreCase);

        merged["Cookie"] = $"token={token}";
        return merged;
    }

    private static string PayloadString(JsonObject payload, string propertyName)
    {
        return payload[propertyName]?.GetValue<string>()
            ?? throw new InvalidOperationException($"Room payload is missing '{propertyName}'.");
    }
}

public sealed record Room(
    [property: JsonPropertyName("roomid")] int RoomId,
    [property: JsonPropertyName("roomName")] string RoomName,
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("accessible")] bool Accessible,
    [property: JsonPropertyName("image")] string Image,
    [property: JsonPropertyName("description")] string Description,
    [property: JsonPropertyName("features")] IReadOnlyList<string> Features,
    [property: JsonPropertyName("roomPrice")] int RoomPrice);

public sealed record Rooms(
    [property: JsonPropertyName("rooms")] IReadOnlyList<Room> Items);
